javatests/arcs/core/util/ParserTest.kt

```diff
@@ -171,11 +171,46 @@ class ParserTest {
             fail()
         }
 
-        val failure = hello("hello!2").map { _, s, e ->
+        val failure = hello("hello!2").map { _, s, e, _ ->
             require(false) { "Shouldn't be called" }
             Failure("", s, e)
         } as Failure
 
         assertThat(failure.error).isEqualTo("Expecting eof")
     }
+
+    object HelloGrammar : Grammar<String>() {
+        val hello by (token("hello") + token("world")).map { (h, w) -> h + w }
+        val world by token("world")
+        val helloOrWorld by hello / world
+        override val topLevel by helloOrWorld + eof
+    }
+
+    @Test
+    fun testGrammar() {
+        val failure = HelloGrammar("foo")
+        assertThat(failure.toString()).isEqualTo("""
+                |Expecting one of hello, world at line 0, column 0
+                |[Traceback]
+                |-->topLevel
+                |---->helloOrWorld
+        """.trimMargin())
+
+        val failure2 = HelloGrammar("hello")
+        assertThat(failure2.toString()).isEqualTo("""
+            |Expecting world at line 0, column 5
```

Comment:
What's your opinion on the order of the stack trace? I am definitely used to JVM stack traces rather then e.g. python ones, that are in the other order.

So I'm wondering whether this would be more readable to folks:

```
Expecting world at line 0, column 5
  at hello
  at helloOrWorld
  at topLevel
```

?